pkg/controller/projects/deploykeys/controller.go

```diff
@@ -0,0 +1,265 @@
+package deploykeys
+
+import (
+	"context"
+	"strconv"
+	"strings"
+
+	compare "github.com/google/go-cmp/cmp"
+	errors "github.com/pkg/errors"
+
+	v1alpha1 "github.com/crossplane-contrib/provider-gitlab/apis/projects/v1alpha1"
+	clients "github.com/crossplane-contrib/provider-gitlab/pkg/clients"
+	projects "github.com/crossplane-contrib/provider-gitlab/pkg/clients/projects"
+	xpv1 "github.com/crossplane/crossplane-runtime/apis/common/v1"
+	event "github.com/crossplane/crossplane-runtime/pkg/event"
+	logging "github.com/crossplane/crossplane-runtime/pkg/logging"
+	meta "github.com/crossplane/crossplane-runtime/pkg/meta"
+	managed "github.com/crossplane/crossplane-runtime/pkg/reconciler/managed"
+	resource "github.com/crossplane/crossplane-runtime/pkg/resource"
+
+	kubeCoreV1 "k8s.io/api/core/v1"
+	kubeTypes "k8s.io/apimachinery/pkg/types"
+	kubeController "sigs.k8s.io/controller-runtime"
+	kubeClient "sigs.k8s.io/controller-runtime/pkg/client"
+
+	gitlab "github.com/xanzy/go-gitlab"
+)
+
+const (
+	ERR_NOT_DEPLOY_KEY = "managed resource is not a Gitlab deploy key custom resource"
+	ERR_NOT_FOUDN      = "404 project deploy key not found"
+	ERR_GET_FAIL       = "cannot get Gitlab deploy key"
+	ERR_CREATE_FAIL    = "cannot create Gitlab deploy key"
+	ERR_UPDATE_FAIL    = "cannot update Gitlab deploy key"
+	ERR_DELETE_FAIL    = "cannot delete Gitlab deploy key"
+	ERR_KEY_MISSING    = "missing key ref value"
+	ERR_ID_NOT_INT     = "external-name is not an int"
```

Comment:
Unlike in other languages, constants or any other kind of variable should not have UPPER_CASE letters. Instead it should be like `errNotDeployKey`.

Btw: Alle Identifiers starting with an upper case letter are considered public. All with a lower case letter are private (package) scope.